### PR TITLE
JS: introduce Expr:getUnderlyingX

### PR DESCRIPTION
**Pull request by esben-semmle**
This PR introduces `Expr::getUnderlyingValue` and its sibling `Expr::getUnderlyingReference` as the semantic versions of what `Expr::stripParens` most often is used for.

I have two unopened PRs that make use of the features of this PR, so I would like it merged soon.

I have another stack of commits that replace most `Expr::stripParens` uses with `Expr::getunderlyingValue` and `Expr:getUnderlyingReference` throughout the code base. But I would like to wait with merging that since the [performance evaluation](https://git.semmle.com/esben/dist-compare-reports/tree/js/syntactic-nullOrUndefined_1539996135170) of that is unstable, and since a timeout has started to consistently happen on the base commit as well.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/TypeScript.qll [hunk 1]
```diff
@@ -1315,6 +1315,15 @@ class TypeAssertion extends Expr, @typeassertion {
   override ControlFlowNode getFirstControlFlowNode() {
     result = getExpression().getFirstControlFlowNode()
   }
+
+  override Expr getUnderlyingValue() {
+    result = getExpression().getUnderlyingValue()
+  }
+
+  override Expr getUnderlyingReference() {
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/TypeScript.qll]: I'm not sure type assertions can be used as LHSs; @asger-semmle?

Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/TypeScript.qll]: Didn't think so either, but it seems to work.

File: javascript/ql/src/semmle/javascript/Expr.qll [hunk 2]
```diff
@@ -45,18 +45,28 @@ class ExprOrType extends @exprortype, Documentable {
...
+  /**
+   * Gets this expression or type, with any surrounding parentheses removed.
+   *
+   * Also see `getUnderlyingValue` and `getUnderlyingReference`.
+   */
   ExprOrType stripParens() { result = this }
 
   /**
    * Gets the innermost reference that this expression evaluates to, if any.
+   * For example, `o.m` of `(o.m)`, but not of `(0, o.m)`.
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/Expr.qll]: I'd prefer a more complete listing of all covered cases (also for `getUnderlyingValue`).

Comment (esben-semmle) [reply javascript/ql/src/semmle/javascript/Expr.qll]: Done.

**Comment by esben-semmle** [comment]
Rebased to resolve expected output conflict.

**Pull request merged**